mu/virtual_environment.py

```diff
@@ -288,12 +285,31 @@ def __init__(self, dirpath=None):
         self._bin_extension = ".exe" if self._is_windows else ""
         self.settings = settings.VirtualEnvironmentSettings()
         self.settings.init()
-        self.relocate(dirpath or self.settings["dirpath"])
+        dirpath_to_use = dirpath or self.settings["dirpath"] or self._generate_dirpath()
+        logger.info("Using dirpath: %s", dirpath_to_use)
+        self.relocate(dirpath_to_use)
 
     def __str__(self):
         return "<%s at %s>" % (self.__class__.__name__, self.path)
 
+    @staticmethod
+    def _generate_dirpath():
+        """Construct a unique virtual environment folder
+
+        To avoid clashing with previously-created virtual environments,
+        construct one which includes the Python version and a timestamp
+        """
+        return "%s-%s-%s" % (

```

Comment:
Just a style issue / question.... is `"%s" % foo` string interpolation supposed to be replaced by the preferred `"{}".format(foo)`..? :shrug:

Reply:
Not especially. It's never been deprecated (and because my use of Python predates the format and fstring approaches I tend to default to the % style). However, if you'd prefer the Mu codebase to standardise on one or another, I'm quite happy to change here and elsewhere.